ide/gototest/src/org/netbeans/api/gototest/TestOppositesLocator.java

```diff
@@ -0,0 +1,194 @@
+
+/**
+ * Find one or multiple test files for a source file,
+ * and one or multiple source files for a test file.
+ *
+ * @since 1.57
+ */
+public class TestOppositesLocator {
+
+    public static TestOppositesLocator getDefault() {
+        return new TestOppositesLocator();
+    }
+
+    /**
+     * Given the file and position in the file, if the:
+     * <ul>
+     *     <li> given file is a source file, find corresponding test file or test files, if exist.</li>
+     *     <li> given file is a test file, find corresponding source file or source files, if exist.</li>
+     * </ul>
+     *
+     * @param fo the file for which the opposites should be found
+     * @param caretOffset position in the file, or {@code -1} if unknown
+     * @return a result describing either an error, or a possibly empty list of locations found;
+     *         note one of {@code errorMessage} and {@code locations} is always {@code null},
+     *         and one always non-{@code null}.
+     */
+    @NbBundle.Messages("No_Test_Or_Tested_Class_Found=No Test or Tested class found")
+    public LocatorResult findOpposites(FileObject fo, int caretOffset) {
+        TestLocator.FileType currentFileType = getCurrentFileType(fo);
+        if(currentFileType == TestLocator.FileType.NEITHER) {
+            return new LocatorResult(Bundle.No_Test_Or_Tested_Class_Found(), null);
+        }
+        else {
+            return new LocatorResult(null, Collections.unmodifiableList(populateLocationResults(fo, caretOffset)));
+        }
+    }
+
+    private List<NamedLocation> populateLocationResults(FileObject fo, int caretOffset) {
+        Map<LocationResult, String> locationResults = new HashMap<LocationResult, String>();
+
+        Collection<? extends TestLocator> locators = Lookup.getDefault()
+                                                           .lookupAll(TestLocator.class)
+                                                           .stream()
+                                                           .filter(tl -> tl.appliesTo(fo))
+                                                           .collect(Collectors.toList());
+
+        CountDownLatch allDone = new CountDownLatch(locators.size());
+
+        for (TestLocator locator : locators) {
+            if (locator.appliesTo(fo)) {
+                if (locator.asynchronous()) {
+                    locator.findOpposite(fo, caretOffset, new LocationListener() {
+                        @Override
+                        public void foundLocation(FileObject fo, TestLocator.LocationResult location) {
+                            if (location != null) {
+                                FileObject fileObject = location.getFileObject();
+                                if(fileObject == null) {
+                                    String msg = location.getErrorMessage();
+                                    if (msg != null) {
+                                        DialogDisplayer.getDefault().notify(
```

Comment:
I think this is problematic. An interactive message should be at least controlled by a flag in the API call (display or just report). Errors should be reported to the client from the API. One never knows when some post-processing, or different presentation style will be needed.